### PR TITLE
Add docs build job

### PR DESCRIPTION
**Pull request by AyodeAwe**
The PR adds a `docs_build` process to the PR and Build workflows for this repository. The generated docs are synced to s3 for only the build workflows.

### CODE REVIEW

**Comment by AyodeAwe** [comment]
Waiting for feedback regarding the java docs build from @bdice

File: ci/build_docs.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+rapids-logger "Downloading artifacts from previous jobs"
+CPP_CHANNEL=$(rapids-download-conda-from-s3 cpp)
+PYTHON_CHANNEL=$(rapids-download-conda-from-s3 python)
+VERSION_NUMBER=$(rapids-get-rapids-version-from-git)
+
+rapids-mamba-retry install \
+  --channel "${CPP_CHANNEL}" \
+  --channel "${PYTHON_CHANNEL}" \
+  libcudf \
+  cudf
```
Comment (ajschmidt8) [inline ci/build_docs.sh:28]:
```suggestion
  libcudf cudf
```

File: .github/workflows/pr.yaml [hunk 2]
```diff
@@ -82,6 +83,16 @@ jobs:
       arch: "amd64"
       container_image: "rapidsai/ci:latest"
       run_script: "ci/test_notebooks.sh"
+  docs-build:
+    needs: conda-python-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@cuda-118
```
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:89]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -53,6 +53,17 @@ jobs:
       date: ${{ inputs.date }}
       sha: ${{ inputs.sha }}
       skip_upload_pkgs: libcudf-example
+  docs-build:
+    if: ${{ startsWith(github.ref, 'refs/heads/branch-') }}
+    needs: python-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@cuda-118
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:60]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

**Comment by bdice** [comment]
> Waiting for feedback regarding the java docs build from @bdice

(Documenting conversation notes from Slack) I tried building Java docs locally a while ago with `mvn build` but I didn't figure it out. @GaryShen2008 may be able to help automate this.

**Comment by ajschmidt8** [comment]
> > Waiting for feedback regarding the java docs build from @bdice
> 
> (Documenting conversation notes from Slack) I tried building Java docs locally a while ago with `mvn build` but I didn't figure it out. @GaryShen2008 may be able to help automate this.

We're going to defer the java docs builds to a future PR so we can get this merged.

**Review by ajschmidt8 — APPROVED** [review]

File: ci/build_docs.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+rapids-logger "Build Sphinx docs"
+pushd docs/cudf
+sphinx-build -b dirhtml source _html
+sphinx-build -b text source _text
+popd
+
+
+if [[ ${RAPIDS_BUILD_TYPE} == "branch" ]]; then
+  aws s3 sync --delete cpp/doxygen/html "s3://rapidsai-docs/cudf/${VERSION_NUMBER}/html"
```
Comment (bdice) [inline ci/build_docs.sh:44]: This looks like a typo with the next line:
```suggestion
  aws s3 sync --delete cpp/doxygen/html "s3://rapidsai-docs/libcudf/${VERSION_NUMBER}/html"
```

File: ci/build_docs.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  aws s3 sync --delete docs/cudf/_html "s3://rapidsai-docs/libcudf/${VERSION_NUMBER}/html"
```
Comment (bdice) [inline ci/build_docs.sh:45]: Typo with previous line?
```suggestion
  aws s3 sync --delete docs/cudf/_html "s3://rapidsai-docs/cudf/${VERSION_NUMBER}/html"
```

File: dependencies.yaml [hunk 6]
```diff
@@ -41,6 +41,12 @@ files:
     includes:
       - develop
       - py_version
+  docs:
+    output: none
+    includes:
```
Comment (bdice) [inline dependencies.yaml:46]: Have we verified that all the notebooks rendered during our docs build can execute with only the runtime dependencies of cudf? (Reminder to self, double check this.)

File: docs/cudf/source/conf.py [hunk 7]
```diff
@@ -23,11 +22,6 @@
 from docutils.nodes import Text
 from sphinx.addnodes import pending_xref
 
-import cudf
```
Comment (bdice) [inline docs/cudf/source/conf.py:26]: How are the docs built without importing the cudf module? I thought that was required for autodoc and other Sphinx features to work but maybe I’m wrong.

**Review by bdice — CHANGES_REQUESTED** [review]
Requesting changes for a typo fix in upload script. I’d like to investigate a couple other questions as well before merging.

File: docs/cudf/source/conf.py (hunk 7 above)
Comment (AyodeAwe) [reply docs/cudf/source/conf.py:26]: [These logs](https://github.com/rapidsai/cudf/actions/runs/4237425138/jobs/7364044147) suggest that the docs build successfully without the `cudf` import.

Comment (ajschmidt8) [reply docs/cudf/source/conf.py:26]: See my commit message in the commit below for additional details

- https://github.com/rapidsai/cudf/pull/12592/commits/70baa0ffb9a495730068d10d4c3dbecbe8565d76

Comment (bdice) [reply docs/cudf/source/conf.py:26]: Sweet. Thanks for verifying!

File: dependencies.yaml (hunk 6 above)
Comment (AyodeAwe) [reply dependencies.yaml:46]: The notebooks render correctly.

**Review by bdice — APPROVED** [review]

**Pull request merged**